File: Lab2/primes/Primes.java
interface Filter {
    public boolean accept(int n);
}

class AcceptFilter implements Filter {
    public AcceptFilter() {}

    public boolean accept(int n) {
        return true;
    }
}

class DivisibleByFilter implements Filter {
    private final int number;
    private final Filter next;

    public DivisibleByFilter(int number, Filter next) {
        this.number = number;
        this.next = next;
    }

    public boolean accept(int n) {
        Filter filter = this;
        while(filter != null && filter instanceof DivisibleByFilter) {
            if(n % ((DivisibleByFilter) filter).number == 0)
                return false;
            filter = ((DivisibleByFilter) filter).next;
        }
        return true;
    }
}

class PrimeChecker {
    private int number;
    private Filter filter;

    public PrimeChecker() {
        this.number = 2;
        this.filter = new AcceptFilter();
    }

    public int next() {
        while (!this.filter.accept(this.number))
            this.number += 1;
        this.filter = new DivisibleByFilter(this.number, this.filter);
        return this.number;
    }

}

public class Primes {

    public static void main(String[] args) {
        PrimeChecker checker = new PrimeChecker();
        int[] primes = new int[5000];
        int count = 0;
        for(int i = 0; i < 5000; i++)
            primes[i] = checker.next();
        System.out.printf("Computed %d prime numbers\n", primes.length);
        System.out.print("The last five are: ");
        for(int i = 4994; i < 5000; i++)
            System.out.printf("%d ", primes[i]);
    }
}

File: Lab2/primes/primes.kt
internal interface Filter {
    fun accept(n: Int): Boolean
}

internal class AcceptFilter : Filter {
    override fun accept(n: Int): Boolean {
        return true
    }
}

internal class DivisibleByFilter(private val number: Int, private val next: Filter) : Filter {
    override fun accept(n: Int): Boolean {
        var filter: Filter = this
        while (filter != null && filter is DivisibleByFilter) {
            if (n % filter.number == 0) return false
            filter = filter.next
        }
        return true
    }
}

internal class PrimeChecker {
    private var number = 2
    private var filter: Filter
    operator fun next(): Int {
        while (!filter.accept(number)) number += 1
        filter = DivisibleByFilter(number, filter)
        return number
    }

    init {
        filter = AcceptFilter()
    }
}


fun main(args: Array<String>) {
    val checker = PrimeChecker()
    val primes = IntArray(5000)
    val count = 0
    for (i in 0..4999) primes[i] = checker.next()
    println("Computed " + primes.size + " prime numbers")
    print("The last five are: ")
    for (i in 4994..4999) print("" + primes[i] + " ")
}
